app/src/main/java/com/github/se/polyfit/ui/screen/MapScreen.kt

```diff
@@ -24,88 +25,112 @@ import androidx.compose.ui.unit.dp
+
   val cameraPositionState = rememberCameraPositionState {
     position = CameraPosition.fromLatLngZoom(LatLng(EPFL_LATITUDE, EPFL_LONGITUDE), 15f)
   }
-  var sliderPosition by remember { mutableFloatStateOf((MIN_SIZE_RAYON + MAX_SIZE_RAYON) / 2) }
-
-  LaunchedEffect(currentLocation) {
-    mutex.withLock {
-      currentLocation.value = mapviewMap.getCurrentLocation()
-      mapviewMap.setLocation(currentLocation.value)
-    }
-
-    mapviewMap.listenToPosts()
-    mapviewMap.posts.observeForever { posts.value = it }
-  }
-  LaunchedEffect(sliderPosition) {
-    isCircleShown = true
-    mutex.withLock { mapviewMap.setRadius(sliderPosition.toDouble() / 1000) }
 
-    if (isCircleShown) {
-      mapviewMap.listenToPosts()
-      mapviewMap.posts.observeForever { posts.value = it }
-    }
-    delay(3000)
-    isCircleShown = false
+  var sliderPositionInMeters by remember {
+    mutableFloatStateOf((MIN_SIZE_RAYON + MAX_SIZE_RAYON) / 2)
   }
 
-  LaunchedEffect(key1 = mapviewMap) {
-    //        mapviewMap.getAllPost().collect { posts.value = it }
-
-    Log.i("Map", "MapScreen $posts")
-    mutex.withLock {
-      currentLocation.value = mapviewMap.getCurrentLocation()
-      mapviewMap.setLocation(currentLocation.value)
-    }
-    Log.i("Map", "currentLocation $currentLocation")
+  /**
+   * Effect to be launched when the mapViewModel changes. Initializes the current location and sets
+   * up the map and post listeners.
+   *
+   * This effect fetches the current location asynchronously, sets the location in the mapViewModel,
+   * and begins listening to posts. It also updates the camera position of the map and observes
+   * posts to update the state.
+   *
+   * @param mapViewModel The ViewModel that handles map-related data and operations.
+   */
+  LaunchedEffect(key1 = mapViewModel) {
+    currentLocation.value = mapViewModel.getCurrentLocation().await()
+    mapViewModel.setLocation(currentLocation.value)
+
+    mapViewModel.listenToPosts()
+    mapViewModel.posts.observeForever { posts.value = it }
 
     cameraPositionState.position =
         CameraPosition.fromLatLngZoom(
             LatLng(currentLocation.value.latitude, currentLocation.value.longitude), 15f)
+
+    mainLaunchedIsFinished = true
+  }
+
+  /**
+   * Handles changes in the slider position. Updates the slider position and adjusts the map's
+   * search radius accordingly. If the main process has finished, it sets the new radius and
+   * triggers an update to listen for posts.
+   *
+   * @param scope The CoroutineScope in which to launch the coroutine.
+   * @param newPosition The new position of the slider.
+   */
+  val mutex = Mutex()
+  fun onSliderChanged(scope: CoroutineScope) {
+    scope.launch {
+      mutex.withLock {
+        if (mainLaunchedIsFinished) {
+
+          mapViewModel.setRadius(sliderPositionInMeters.toDouble() / 1000)
+          mapViewModel.listenToPosts()
+          delay(THREE_SECONDS)
```

Comment:
In general should not use delays in the code, but this time it is necessary. I would add a comment to say something along the line that it is need to hide the circle radius after a certain time

Reply:
how do you make a delay without using delay?